### PR TITLE
Added a page documenting scikit-learn for Aurora

### PR DESCRIPTION
**Pull request by BethanyL**

### CODE REVIEW

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+from sklearnex.neighbors import NearestNeighbors
+```
+
+### GPU Acceleration
+Intel Extension for Scikit-learn can execute algorithms on the GPU via the [dpctl](https://intelpython.github.io/dpctl/latest/index.html) package, which should be included in the frameworks module. (If not, see the [Python page](../python.md)) dpctl implements oneAPI concepts like queues and devices. 
+
+As described in more detail in Intel's documentation [here](https://uxlfoundation.github.io/scikit-learn-intelex/latest/oneapi-gpu.html), there are two ways to run on the GPU. 
+
+1. Pass the input data to the algorithm as `dpctl.tensor.usm_ndarray`. Then the algorithm will run on the same device as the data and return the result as a usm_array on the same device. 
+2. Configure Intel Extension for Scikit-learn, for example, by setting a context: `sklearnex.config_context`.
```
Comment (felker) [inline docs/aurora/data-science/frameworks/scikit-learn.md:34]:
```suggestion
2. Configure Intel Extension for scikit-learn, for example, by setting a context: `sklearnex.config_context`. 
```

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+### GPU Acceleration
+Intel Extension for Scikit-learn can execute algorithms on the GPU via the [dpctl](https://intelpython.github.io/dpctl/latest/index.html) package, which should be included in the frameworks module. (If not, see the [Python page](../python.md)) dpctl implements oneAPI concepts like queues and devices. 
+
+As described in more detail in Intel's documentation [here](https://uxlfoundation.github.io/scikit-learn-intelex/latest/oneapi-gpu.html), there are two ways to run on the GPU. 
+
+1. Pass the input data to the algorithm as `dpctl.tensor.usm_ndarray`. Then the algorithm will run on the same device as the data and return the result as a usm_array on the same device. 
+2. Configure Intel Extension for Scikit-learn, for example, by setting a context: `sklearnex.config_context`. 
+
+### Distributed Mode
+To distribute an sklearnex algorithm across multiple GPUs, we need several ingredients, which we will demonstrate in an example below. We recommend using the MPI backend rather than the CCL backend since it is more tested on Aurora.
```
Comment (felker) [inline docs/aurora/data-science/frameworks/scikit-learn.md:37]:
```suggestion
To distribute an `sklearnex` algorithm across multiple GPUs, we need several ingredients demonstrated in an example below. We recommend using the MPI backend rather than the CCL backend since it is more tested on Aurora.
```

"more tested"  phrasing sounds awkward to me. Maybe "since it is tested more thoroughly on Aurora."

**Review by felker — CHANGES_REQUESTED** [review]
just an observation: we don't have a scikit-learn page in any other Machine section. are there architecture/platform-specific performance notes we want to convey for scikit-learn users on Polaris, Sophia, etc? Or is it all straightforward with cuML integration on NVIDIA platforms. What about scaling?

**Comment by BethanyL** [comment]
> just an observation: we don't have a scikit-learn page in any other Machine section. are there architecture/platform-specific performance notes we want to convey for scikit-learn users on Polaris, Sophia, etc? Or is it all straightforward with cuML integration on NVIDIA platforms. What about scaling?

Basically scikit-learn doesn't run on NVIDIA GPUs, but cuML has similar functionality and APIs. We could add documentation for cuML on Polaris & Sophia in the future if we have advice specific to our platforms or scaling.

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+let NRANKS=${NNODES}*${NRANKS_PER_NODE}
+
+#####################################################################
+# Environment set up, using the latest frameworks drop
+#####################################################################
+
+module use /soft/modulefiles
+module load frameworks
+
+export NUMEXPR_NUM_THREADS=64
```
[the remark this answers is not in the document]
Comment (BethanyL) [reply docs/aurora/data-science/frameworks/scikit-learn.md:110]: I know at least before I was getting a warning if I didn't set this, but maybe the warning disappears for distributed runs because then each rank has fewer threads. I can check. The issue is with a package called [numexpr](https://github.com/pydata/numexpr), which sets numexpr.numthreads to the available number of threads by default, which on an Aurora node is 208. However, NUMEXPR_MAX_THREADS is set to 64 as a default. The solution is to either set NUMEXPR_NUM_THREADS to <= 64 or increase NUMEXPR_MAX_THREADS.

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+#####################################################################
+# JOB LAUNCH
+######################################################################
+
+export CPU_BIND="verbose,list:2-4:10-12:18-20:26-28:34-36:42-44:54-56:62-64:70-72:78-80:86-88:94-96"
+
+# Launch the script
+mpiexec -np ${NRANKS} -ppn ${NRANKS_PER_NODE} \
+--cpu-bind ${CPU_BIND} gpu_tile_compact.sh \
```
[the remark this answers is not in the document]
Comment (BethanyL) [reply docs/aurora/data-science/frameworks/scikit-learn.md:131]: Oh good question. On Aurora, mpiexec defaults to --hostfile $PBS_NODEFILE if --hostfile is not included.

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+module. You can load the frameworks module as described [here](../python.md), which will activate a conda environment. 
+
+## Usage
+### Patching
+There are multiple ways to patch scikit-learn with the Intel Extension, as Intel documents [here](https://uxlfoundation.github.io/scikit-learn-intelex/latest/what-is-patching.html). For example, you can patch within the script, like this:
+```
+from sklearnex import patch_sklearn
+patch_sklearn()
+```
+It is important to note that this needs to happen before importing scikit-learn. To explicitly only patch certain estimators, you can import particular functions from sklearnex instead of sklearn, like this:
```
[the remark this answers is not in the document]
Comment (BethanyL) [reply docs/aurora/data-science/frameworks/scikit-learn.md]: Thanks! I didn't think about it that way. I'm updating the wording.

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+To distribute an sklearnex algorithm across multiple GPUs, we need several ingredients, which we will demonstrate in an example below. We recommend using the MPI backend rather than the CCL backend since it is more tested on Aurora.
+
+!!! warning "Warning"
+    The current version of Intel Extension to scikit-learn does not scale well to multiple GPUs. The cause has been identified, and we're waiting on a fix. However, if you use the oneDAL C++ API, the scaling is much better.
+
+1. Create an MPI communicator using mpi4py if you need to use the rank. (mpi4py is also included in the frameworks module.)
+2. Check for GPU devices. 
+3. Use dpctl to create a SYCL queue (connection to the GPU devices you choose).
+4. Using dpctl and your queue, move your data to the GPU devices.
+5. Run the algorithm on that data. The compute will happen where the data is. The algorithm should be from `sklearnex.spmd`.
```
[the remark this answers is not in the document]
Comment (BethanyL) [reply docs/aurora/data-science/frameworks/scikit-learn.md:46]: Oh yes, that's more clear, thank you!

File: docs/aurora/data-science/frameworks/scikit-learn.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+# Run the algorithm. 
+model_spmd = KNeighborsClassifier(
+    algorithm="brute", n_neighbors=20, weights="uniform", p=2, metric="minkowski"
+)
+model_spmd.fit(dpt_X, dpt_y)
+```
+
+### An Example Job Script
+
+Below we give an example job script, adapted from [the PyTorch page](pytorch.md). Note that we are using Aurora MPICH (the default MPI library on Aurora) and not using oneCCL, so we don't need special oneCCL settings. For more about pinning ranks to CPU cores and GPUs, see the [Running Jobs page](../../running-jobs-aurora.md).
```
[the remark this answers is not in the document]
Comment (felker) [reply docs/aurora/data-science/frameworks/scikit-learn.md:82]: seems like the whole scikit-learn-intelex/Aurora-specific purpose of this example script is to discuss the `numexpr` workaround--- might be better to remove the example script and just discuss those lines on their own, linking to other generic example scripts? either way is fine. I added line highlights for that section if you want to keep the script as-is

**Review by felker — CHANGES_REQUESTED** [review]

**Comment by felker** [comment]
strange, none of the suggested changes seemed to make it to the final version via 4f27a310f745d828ef22d76d30f2389cba2fc851

I will clean up the code blocks in a separate PR
